## <center>Case Study on Probability for Data Science

### 1) Reading the dataset to python environment

In [1]:
import pandas as pd
mushroom=pd.read_csv('mushrooms.csv')

In [2]:
#Checking the overall structure of the data set
mushroom

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

The dataset consist of 8124 rows and 23 columns

### 2) Pre-processing

In [3]:
#Checking for null values

In [4]:
mushroom.isna().sum()

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

Dataset doesnt contain any null values

In [5]:
# Using label encoding to convert the column data to numerical type

In [6]:
M=mushroom.copy() # To be used for Model building with one hot encoded data

In [7]:
from sklearn.preprocessing import LabelEncoder
label=LabelEncoder()
for col in mushroom.columns:
    mushroom[col] = label.fit_transform(mushroom[col])

In [8]:
mushroom

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-below-ring  stalk-color-above-ring  \
0                            2                       7   
1                            2                       7   
2                            2                       7   
3                            2                       7   
4                            2                       7   
...                        ...                     ...   
8119                         2                       5   
8120                         2                       5   
8121                         2                       5   
8122                         1                       7   
8123                         2                       5   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
0                          7          0           2            1          4   
1                          7          0           2            1          4   
2                          7          0           2            1          4   
3                          7          0           2            1          4   
4                          7          0           2            1          0   
...                      ...        ...         ...          ...        ...   
8119                       5          0           1            1          4   
8120                       5          0           0            1          4   
8121                       5          0           1            1          4   
8122                       7          0           2            1          0   
8123                       5          0           1            1          4   

      spore-print-color  population  habitat  
0                     2           3        5  
1                     3           2        1  
2                     3           2        3  
3                     2           3        5  
4                     3           0        1  
...                 ...         ...      ...  
8119                  0           1        2  
8120                  0           4        2  
8121                  0           1        2  
8122                  7           4        2  
8123                  4           1        2  

[8124 rows x 23 columns]

In [9]:
#Checking for unique values in each of the columns.

In [10]:
for col in mushroom.columns:
    print(col,mushroom[col].unique())

class [1 0]
cap-shape [5 0 4 2 3 1]
cap-surface [2 3 0 1]
cap-color [4 9 8 3 2 5 0 7 1 6]
bruises [1 0]
odor [6 0 3 5 2 1 8 7 4]
gill-attachment [1 0]
gill-spacing [0 1]
gill-size [1 0]
gill-color [ 4  5  2  7 10  3  9  1  0  8 11  6]
stalk-shape [0 1]
stalk-root [3 2 1 4 0]
stalk-surface-above-ring [2 0 1 3]
stalk-surface-below-ring [2 0 3 1]
stalk-color-above-ring [7 3 6 4 0 2 5 1 8]
stalk-color-below-ring [7 6 3 0 4 2 8 5 1]
veil-type [0]
veil-color [2 0 1 3]
ring-number [1 2 0]
ring-type [4 0 2 1 3]
spore-print-color [2 3 6 1 7 5 4 8 0]
population [3 2 0 4 5 1]
habitat [5 1 3 0 4 6 2]


veil-type has got only one value,so the column can be dropped,since its not relevant in training the model.

In [11]:
mushroom.drop('veil-type',inplace=True,axis=1)

### 3) Building the models

In [12]:
#class is the target and to know the count of each class using value_counts()
mushroom['class'].value_counts(normalize=True)  # using normalize to get the proportion

0    0.517971
1    0.482029
Name: class, dtype: float64

Class distribution is almost equal

In [13]:
# spliiting the dataset into features and target
#taking 'Class' as target(y) and all other as X
X=mushroom.drop(['class'],axis=1)
y=mushroom['class']

In [14]:
# splitting into test and training set
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.20)

In [15]:
# checking how many data points are there in the training set
X_train.shape

(6499, 21)

In [16]:
# checking how many data points are there in the training set
X_test.shape

(1625, 21)

There are 6499 data points for training and 1625 data points for testing

#### 3.1) Logistics regression

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,f1_score
logit_model=LogisticRegression(random_state=42,max_iter=1000)
logit_model.fit(X_train,y_train)
y_pred=logit_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 0.948
f1 score is: 0.946


#### 3.2) SVM Classifier

In [18]:
from sklearn.svm import SVC
svm_model=SVC(random_state=42)
svm_model.fit(X_train,y_train)
y_pred=svm_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 0.994
f1 score is: 0.994


#### 3.3) KNN Classifier

In [19]:
from sklearn.neighbors import KNeighborsClassifier
knn_model=KNeighborsClassifier()
knn_model.fit(X_train,y_train)
y_pred=knn_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 0.996
f1 score is: 0.996


#### 3.4)  Decision tree Classifier

In [20]:
from sklearn.tree import DecisionTreeClassifier
decision_tree=DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train,y_train)
y_pred=decision_tree.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 1.0
f1 score is: 1.0


#### 3.5) Random forest Classifier 

In [21]:
from sklearn.ensemble import RandomForestClassifier
random_forest=RandomForestClassifier(random_state=42)
random_forest.fit(X_train,y_train)
y_pred=random_forest.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))

Accuracy is: 1.0
f1 score is: 1.0


#### 3.6) Categorical NB

In [22]:
from  sklearn.naive_bayes import CategoricalNB
cat_NB=CategoricalNB()
cat_NB.fit(X_train,y_train)
y_pred=cat_NB.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))

Accuracy is: 0.951
f1 score is: 0.95


Categorical NB has been used since the features are categorical.
Bernoulli NB can be applied after one hot encoding,but one hot encoded columns will be highly correlated which is against the assumptions of Naive Bayes classifier. 
Bernoulli NB can be applied in label encoded data where the model itself binarize the data ,by assuming all non zero categories as 1.
Gaussian NB is not applied since ,the features are not continuous data.

#### 3.7)Bernoullli NB(with label encoding)

In [23]:
#Applying Bernoullli NB to check how it performs.

In [24]:
from sklearn.naive_bayes import BernoulliNB
B_NB=BernoulliNB()
B_NB.fit(X_train,y_train)
y_pred=B_NB.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))

Accuracy is: 0.842
f1 score is: 0.82


#### 3.8) Bernoullli NB(with onehot encoding)

In [41]:
# Applying Bernoullli NB after onehot encoding

In [27]:
X_df=M.drop('class',axis=1)
y=M['class']

In [39]:
#One hot encoding using get dummies
X=pd.get_dummies(X_df)  #Not encoding the target,since scikit learn automatically encodes the target
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.20)
from sklearn.naive_bayes import BernoulliNB
B_NB=BernoulliNB()
B_NB.fit(X_train,y_train)
y_pred=B_NB.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))


Accuracy is: 0.936
f1 score is: 0.93


### 4) Evaluation of Models

#### 4.1) Evaluation of Model(with Label encoded data)

| Model | Accuracy | F1_score
| --- | --- | --- |
| Logistic regression | 0.948 | 0.946 
| SVM | 0.993 | 0.992 
| KNN | 0.996 | 0.996  
| Decision Tree |1.0 | 1.0 
| Random Forest | 1.0 | 1.0
| CategoricalNB | 0.951 | 0.951
| BernoulliNB(label encoding) | 0.842 | 0.82
|BernoulliNB(one hot encoding) | 0.936 | 0.93

Decision tree & Random forest classifier gives the best result.
In the case of Naive Bayes classifier,CategoricalNB gives the best result.Bernoulli NB performance improved with onehot encoding encoding.

In [ ]:
#Since BernoulliNB improved with onehot encoding ,checking whether performance of other models improves with onehot encoded data

#### 4.2)Model Building (with onehot encoding) 

In [51]:
# Testing whether Logistic regression,SVM and KNN improves by one hot encoding the features.
# For categorical NB one hot encoding is not applied.(since it exactly requires categorical features)

##### 4.2.1)Logistic regression

In [47]:
#One hot encoding using get dummies
X=pd.get_dummies(X_df)  #Not encoding the target,since scikit learn automatically encodes the target
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42,test_size=0.20)
logit_model=LogisticRegression(random_state=42,max_iter=1000)
logit_model.fit(X_train,y_train)
y_pred=logit_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))

Accuracy is: 1.0
f1 score is: 1.0


##### 4.2.2) SVM Classifier

In [48]:
from sklearn.svm import SVC
svm_model=SVC(random_state=42)
svm_model.fit(X_train,y_train)
y_pred=svm_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 1.0
f1 score is: 1.0


##### 4.2.3) KNN Classifier

In [50]:
from sklearn.neighbors import KNeighborsClassifier
knn_model=KNeighborsClassifier()
knn_model.fit(X_train,y_train)
y_pred=knn_model.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),3))

Accuracy is: 1.0
f1 score is: 1.0


In [54]:
#Tried CategoricalNB with one hot enoded data
from  sklearn.naive_bayes import CategoricalNB
cat_NB=CategoricalNB()
cat_NB.fit(X_train,y_train)
y_pred=cat_NB.predict(X_test)
print("Accuracy is:",round(accuracy_score(y_test,y_pred),3))
print("f1 score is:",round(f1_score(y_test,y_pred),2))

Accuracy is: 0.936
f1 score is: 0.93


In [55]:
#Result is better with label encoded data

Logistic regression,SVM,KNN models improved after onehot encoding the features

| Model | Accuracy | F1_score 
| --- | --- | --- |
| Logistic regression(Label encoding) | 0.948 | 0.946 
|Logistic regression(onehot encoding) | 1.0 | 1.0
| SVM(Label encoding) | 0.993 | 0.992
| SVM(one hot encoding) | 1.0 | 1.0
| KNN(Label encoding| 0.996 | 0.996 
| KNN(One hot encoding) | 1.0 | 1.0
| Decision Tree(Label encoding) |1.0 | 1.0 
| Random Forest(Label encoding) | 1.0 | 1.0
| CategoricalNB(Label encoding) | 0.951 | 0.951
| BernoulliNB(label encoding) | 0.842 | 0.82
|BernoulliNB(one hot encoding) | 0.936 | 0.93

Only with One hot encoded data,Logistic regression,SVM and KNN gave 100% accuracy

Decision tree and Random forest classifiers gave 100% accuracy even with label encoded data.

With respect to Naive Bayes classifier,Categorical NB gave the optimum result 